rmw_dds_common/src/graph_cache.cpp

```diff
@@ -0,0 +1,704 @@
+  node_info.node_name = node_name;
+  node_info.node_namespace = node_namespace;
+  it->second.emplace_back(node_info);
+
+  return __create_participant_info_message(participant_gid, it->second);
+}
+
+rmw_dds_common::msg::ParticipantEntitiesInfo
+GraphCache::remove_node(
+  const rmw_gid_t & participant_gid,
+  const std::string & node_name,
+  const std::string & node_namespace)
+{
+  auto it = participants_.find(participant_gid);
+  assert(it != participants_.end());
+
+  auto new_end = std::remove_if(
+    it->second.begin(),
+    it->second.end(),
+    [&](const rmw_dds_common::msg::NodeEntitiesInfo & node_info) {
+      return node_info.node_name == node_name && node_info.node_namespace == node_namespace;
+    });
+
+  assert(new_end != it->second.end());
+
+  it->second.erase(new_end, it->second.end());
+
+  return __create_participant_info_message(participant_gid, it->second);
+}
+
+template<typename FunctorT>
+rmw_dds_common::msg::ParticipantEntitiesInfo
+__modify_node_info(
+  const rmw_gid_t & participant_gid,
+  const std::string & node_name,
+  const std::string & node_namespace,
+  FunctorT do_sth,
+  GraphCache::ParticipantToNodesMap & participant_map)
+{
+  auto participant_info = participant_map.find(participant_gid);
+  assert(participant_info != participant_map.end());
+  auto node_info = std::find_if(
+    participant_info->second.begin(),
+    participant_info->second.end(),
+    [&](const rmw_dds_common::msg::NodeEntitiesInfo & node_info)
+    {
+      return node_info.node_name == node_name && node_info.node_namespace == node_namespace;
+    });
+  assert(node_info != participant_info->second.end());
+
+  do_sth(*node_info);
+  return __create_participant_info_message(participant_gid, participant_info->second);
+}
+
+rmw_dds_common::msg::ParticipantEntitiesInfo
+GraphCache::associate_writer(
+  const rmw_gid_t & writer_gid,
+  const rmw_gid_t & participant_gid,
+  const std::string & node_name,
+  const std::string & node_namespace)
+{
+  auto add_writer_gid = [&](rmw_dds_common::msg::NodeEntitiesInfo & info)
+    {
+      info.writer_gid_seq.emplace_back();
+      convert_gid_to_msg(&writer_gid, &info.writer_gid_seq.back());
+    };
+
+  return __modify_node_info(
+    participant_gid, node_name, node_namespace, add_writer_gid, participants_);
+}
+
+rmw_dds_common::msg::ParticipantEntitiesInfo
+GraphCache::dissociate_writer(
+  const rmw_gid_t & writer_gid,
+  const rmw_gid_t & participant_gid,
+  const std::string & node_name,
+  const std::string & node_namespace)
+{
+  rmw_dds_common::msg::Gid writer_gid_msg;
+  convert_gid_to_msg(&writer_gid, &writer_gid_msg);
+  auto delete_writer_gid = [&](rmw_dds_common::msg::NodeEntitiesInfo & info)
+    {
+      info.writer_gid_seq.erase(
+        std::find_if(
+          info.writer_gid_seq.begin(),
+          info.writer_gid_seq.end(),
+          [&](const rmw_dds_common::msg::Gid & gid)
+          {
+            return gid == writer_gid_msg;
```

Comment:
@ivanpauno here and elsewhere, consider the possibility that the entity being dissociated is not found.

Reply:
If it's not found, there is nothing to do.
I can log a warning in that case.

Reply:
If it's not found, `std::find_if` will return `info.writer_gid_seq.end()` but `std::map::erase` [needs a valid a dereferenceable iterator](https://en.cppreference.com/w/cpp/container/map/erase).